### PR TITLE
[Datasets] Use read stage name for naming Data-read tasks on Ray Dashboard

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?
Previously, the name of read tasks uses the generic `DoRead` name:
screen_shot_2023-04-11_at_4 08 22_pm

This PR updates the naming so that we use the underlying read stage name, if available from the input `LazyBlockList`, as the resulting `MapOperator`; otherwise, we fall back to the existing `DoRead` name. For example, when running this code:
```
import ray

ctx = ray.data.DataContext.get_current()
ctx.new_execution_backend = True

ds = ray.data.read_parquet("example://iris.parquet")
data = ds.take(5)
import time
time.sleep(20)
```

This generates the task with the name `ReadParquet`:
Screenshot at Apr 12 14-52-33

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]
Is there a way we can test this?

**Review by c21 — APPROVED** [review]
yeah let's also add a unit test if it's not too hard.

**Comment by pcmoritz** [comment]
Very nice!

File: python/ray/data/tests/test_execution_optimizer.py [hunk 1]
```diff
@@ -1078,6 +1079,18 @@ def test_from_torch_e2e(ray_start_regular_shared, enable_optimizer, tmp_path):
     _check_usage_record(["FromItems"])
 
 
+def test__blocks_to_input_buffer_op_name(
```
Comment (c21) [inline python/ray/data/tests/test_execution_optimizer.py:1082]: seems extra space? `test__blocks` -> `test_blocks`

File: python/ray/data/tests/test_execution_optimizer.py [hunk 2]
```diff
@@ -1078,6 +1079,18 @@ def test_from_torch_e2e(ray_start_regular_shared, enable_optimizer, tmp_path):
...
+    ray_start_regular_shared,
+    enable_optimizer,
```
Comment (c21) [inline python/ray/data/tests/test_execution_optimizer.py:1084]: we don't need `enable_optimizer`, right?

**Review by c21 — APPROVED** [review]

File: python/ray/data/tests/test_execution_optimizer.py (hunk 1 above)
Comment (scottjlee) [reply python/ray/data/tests/test_execution_optimizer.py:1082]: The method name is `_blocks_to_input_buffer` with a `_` prefix, so I have it with a double `__`. Let me know if the typical convention is to condense it to a single `_` though.

Comment (c21) [reply python/ray/data/tests/test_execution_optimizer.py:1082]: yeah let's always use single `_` in test name, thanks.

File: python/ray/data/tests/test_execution_optimizer.py (hunk 2 above)
Comment (scottjlee) [reply python/ray/data/tests/test_execution_optimizer.py]: yeah you're right, removed. thanks

**Comment by c21** [comment]
Documentation failure is not related to this PR. Merging to master.

**Pull request merged**